pkg/xds/server/callbacks/dataplane_lifecycle.go

```diff
@@ -188,14 +201,40 @@ func (d *DataplaneLifecycle) registerZoneEgress(ctx context.Context, ze *core_me
 	})
 }
 
-func (d *DataplaneLifecycle) unregisterDataplane(key model.ResourceKey) error {
-	return d.resManager.Delete(context.Background(), core_mesh.NewDataplaneResource(), store.DeleteBy(key))
+func (d *DataplaneLifecycle) deregisterDataplane(key model.ResourceKey, log logr.Logger) error {
+	return d.deregisterProxy(key, core_mesh.NewDataplaneResource(), core_mesh.NewDataplaneInsightResource(), log)
+}
+
+func (d *DataplaneLifecycle) deregisterZoneIngress(key model.ResourceKey, log logr.Logger) error {
+	return d.deregisterProxy(key, core_mesh.NewZoneIngressResource(), core_mesh.NewZoneIngressInsightResource(), log)
 }
 
-func (d *DataplaneLifecycle) unregisterZoneIngress(key model.ResourceKey) error {
-	return d.resManager.Delete(context.Background(), core_mesh.NewZoneIngressResource(), store.DeleteBy(key))
+func (d *DataplaneLifecycle) deregisterZoneEgress(key model.ResourceKey, log logr.Logger) error {
+	return d.deregisterProxy(key, core_mesh.NewZoneEgressResource(), core_mesh.NewZoneEgressInsightResource(), log)
 }
 
-func (d *DataplaneLifecycle) unregisterZoneEgress(key model.ResourceKey) error {
-	return d.resManager.Delete(context.Background(), core_mesh.NewZoneEgressResource(), store.DeleteBy(key))
+func (d *DataplaneLifecycle) deregisterProxy(
+	key model.ResourceKey,
+	obj model.Resource,
+	insight model.Resource,
+	log logr.Logger,
+) error {
+	log.Info("waiting for deregister proxy", "waitFor", d.deregistrationDelay)
+	time.Sleep(d.deregistrationDelay)
```

Comment:
Ha, actually using context from the stream is wrong, because it's canceled by this point. We should use background context. I added a comment with an explanation but I also keep the context in `OnProxyDisconnected`